Sources/Common/Communication/EventBus.swift

```diff
@@ -1,140 +1,104 @@
-    @discardableResult func onReceive<E: EventRepresentable, S: Scheduler>(_ eventType: E.Type, performOn scheduler: S, action: @escaping (E) -> Void) -> AnyCancellable
-
-    var listenersRegistry: EventListenersRegistry { get }
+public protocol EventBus {
+    @discardableResult func post<E: EventRepresentable>(_ event: E, on queue: DispatchQueue?) -> Bool
+    func addObserver<E: EventRepresentable>(_ eventType: E.Type, action: @escaping (E) -> Void)
+    func removeAllObservers()
+    func removeObserver<E: EventRepresentable>(for eventType: E.Type)
 }
 
-/// `SharedEventBus` is a centralized component that manages event distribution in an application.
-/// It allows for the sending of events and the registration of listeners to handle specific event types.
-///
-/// The EventBus follows a publish-subscribe pattern, enabling loose coupling between event producers and consumers.
-///
-/// Usage:
-/// - To send an event: `eventBus.send(myEvent)`
-/// - To listen for an event: `eventBus.onReceive(MyEventType.self) { event in /* handle event */ }`
-///
-/// Important:
-/// - `SharedEventBus` emits `NewSubscriptionEvent` when a new listener subscribes to a specific event type.
-///   This allows other parts of the system to react to changes in event listeners.
-/// - Be cautious when subscribing to `NewSubscriptionEvent` within its own handler.
-///   Creating a new subscription to `NewSubscriptionEvent` in its handler can lead to unintended recursion
-///   and should be avoided to prevent potential infinite loops.
-///
-/// Thread Safety:
-/// - `SharedEventBus` is designed to be thread-safe. It ensures that events are sent and listeners are registered
-///   in a thread-safe manner.
-///
-/// Example:
-/// ```
-/// let eventBus = SharedEventBus(listenersRegistry: ...)
-/// eventBus.onReceive(ProfileIdentifiedEvent.self) { event in
-///     print("Received ProfileIdentifiedEvent with identifier: \(event.identifier)")
-/// }
-/// eventBus.send(ProfileIdentifiedEvent(identifier: "user123"))
-/// ```
-
+/// `SharedEventBus` manages the distribution of events to registered observers.
+/// It uses the NotificationCenter for event delivery and manages observer lifecycles.
 // sourcery: InjectRegisterShared = "EventBus"
 // sourcery: InjectSingleton
 public class SharedEventBus: EventBus {
-    public var listenersRegistry: EventListenersRegistry
+    private var notificationCenter: NotificationCenter = .default
+    private var observers: [String: [NSObjectProtocol]] = [:]
+    private let queue = DispatchQueue(label: "com.eventbus.shared")
 
-    /// Initializes a new instance of `SharedEventBus`.
-    ///
-    /// - Parameter listenersRegistry: The registry used to manage listeners for different event types.
-    public init(listenersRegistry: EventListenersRegistry) {
-        self.listenersRegistry = listenersRegistry
+    deinit {
+        removeAllObservers()
     }
```

Comment:
I am wondering why this class has it's own observers data structure when the NotificationCenter already stores these observers for us? 

From reading the code, it looks like the reason is to remove all observers from the NotificationCenter when the eventbus `deinit` is called? 

If this is the reason, I don't think this will work. We run the risk of memory leaks by this implementation where classes that observe a NotificationCenter event are kept in memory instead of getting destroyed.

The only way that I can think of to prevent this scenario is to have each class remove itself as the observer in it's own `deinit` function: 

```swift
class Foo {

  ...
  eventBus.addObserver(self, ...)
  ...

  deinit {
    eventBus.removeObserver(self)
  }
}
```

Reply:
I don't currently understand the need for a `DispatchQueue` in this class.

From looking at the code, it seems like it's being used to create a mutex around the `observers` property. Is this correct?

Reply:
> NotificationCenter, doesn't provide the information of hasObserver, so if we want to check if there any observers for any perticular event, we can't do so. Hence we are keeping a wrapper around notification center. This is needed for event storage, e.g if there is no observer around for tracking metric, we would know and hence persist the event.

Thanks that's helpful. 

I want to avoid keeping strong references to observers to avoid possible memory leaks. I prefer we let the NotificationCenter be the only one that keeps references to the observer objects. 

Could we change the implementation to keeping an `Int` count of the number of observers instead?

Reply:
> For thread-safety.

The implementation of the `DispatchQueue` looks to be the same as if the `@Atomic` property wrapper was used instead. Does `@Atomic` not work in this case?